jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/AsyncContent.java

```diff
@@ -50,7 +50,31 @@ public class AsyncContent implements Content.Sink, Content.Source, Closeable
     @Override
     public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
     {
-        write(Content.Chunk.from(byteBuffer, last, callback::succeeded), callback);
+        // This class' read() impl calls succeeded() on the callback before returning it to the
+        // caller, so when EOF / EMPTY are used this is perfectly right as the chunk can be
+        // considered consumed and there is no buffer to reclaim. But when a proper buffer is
+        // used the passed callback must be wrapped to make its succeeded() method a noop
+        // as it is the release of the chunk that should trigger the success of the callback.
```

Comment:
I'd change this comment to a javadoc to explain the semantic of `write()` which is: "the callback is completed when the chunk is read *and* released".

Reply:
Done, but handling EOF/EMPTY adds some complexity which I accounted for in the javadoc.